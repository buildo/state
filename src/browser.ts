import createHistory from 'history/createBrowserHistory';
import trim = require('lodash/trim');
import omit = require('lodash/omit');
import * as qs from 'qs';

export type BrowserState = { [k: string]: string } & { view: string };

// TODO: makes it hardly testable and also not configurable for iso (needs to be static location + not a singleton)
const history = createHistory();

// TODO: below, ignoring path params for the moment
// there's only `view` which is the whole `pathname` from history perspective

export function syncToBrowser(s: BrowserState, push: boolean = true): void {
  const pathname = `/${s.view}`;
  const search = `?${qs.stringify(omit(s, ['view']))}`;
  (push ? history.push : history.replace)({ pathname, search });
}

function parseBrowserState(location: typeof history.location): BrowserState {
  const view = trim(location.pathname, ' /');
  const s = qs.parse(location.search);
  return { ...s, view };
}

export function onBrowserChange(callback: (s: BrowserState) => void): void {
  history.listen(location => {
    setTimeout(() => {
      // TODO: something better than setTimeout here?
      callback(parseBrowserState(location));
    });
  });
  setTimeout(() => {
    // TODO: something better than setTimeout here?
    callback(parseBrowserState(history.location));
  });
}
