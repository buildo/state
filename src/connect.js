import React from 'react';
import debug from 'debug';
import { filter } from 'rxjs/operator/filter';
import { map } from 'rxjs/operator/map';
import omit from 'lodash/omit';
import identity from 'lodash/identity';
import pick from 'lodash/pick';
import t from 'tcomb';
import shallowEqual from './shallowEqual';

const warn = debug('state:connect');
warn.log = ::console.warn; // eslint-disable-line no-console

export const ConnectContextTypes = {
  transition: React.PropTypes.func.isRequired,
  state: React.PropTypes.object.isRequired
};

const defaultGetNewState = ty => v => ty.reduce((acc, key) => ({
  ...acc,
  [key]: v[key]
}), {});

// expects a select function or a list of keys
// and an optional configuration object
export default stateType => (select = identity, {
  // implement a standard shouldComponentUpdate with shallowEquals
  // do not use on non-pure components, e.g. react-router `RouteHandler`s
  //
  // Boolean
  //
  pure = true,

  // do not update the component props if the type declaration doesn't match
  // useful to skip "router transitioning" problems: old container that should be unmounted
  // soon (but still async) by re-rendering after a router change would otherwise receive
  // the new (and possibly incorrect from their POV) state
  // TODO(gio): elaborate better / try to remove
  //
  // Boolean
  //
  filterValid = true,

  // optionally kill some props using lodash omit
  // useful to stay strict in @props declaration, a typical default could be:
  // ['router', 'query', 'params']
  //
  // Array<String>
  //
  killProps = []
} = {}) => {

  const isKeyList = t.list(t.String).is(select);
  if (!isKeyList && !t.Function.is(select)) {
    throw new Error('connect expects a select function or a list of keys');
  }

  select.forEach(k => {
    if (!stateType.meta.props.hasOwnProperty(k)) {
      throw new Error(`${k} is not defined in state`);
    }
  });

  const decorator = Component => {
    const displayName = `connect(${Component.displayName || Component.name || 'Component'})`;

    const getNewState = isKeyList ? defaultGetNewState(select) : select;

    const shouldUpdateState = filterValid && isKeyList ? v => {
      const invalid = select.reduce((acc, k) => {
        return !stateType.meta.props[k].is(v[k]) ? [...acc, k] : acc;
      }, []);

      if (invalid.length > 0 && process.env.NODE_ENV === 'development') {
        warn(`Skipping update for ${displayName}. Invalid keys: ${invalid.join(', ')}`);
      }
      return invalid.length === 0;
    } : () => true;

    return class ConnectWrapper extends React.Component {
      static contextTypes = ConnectContextTypes

      static displayName = displayName;

      constructor(props, context) {
        super(props, context);
        const value = context.state.value;
        if (shouldUpdateState(value)) {
          this.state = getNewState(value);
        } else {
          this.state = {};
        }
      }

      componentDidMount() {
        this._subscription = this.context.state::filter(shouldUpdateState)::map(getNewState).subscribe(::this.setState);
      }

      componentWillUnmount() {
        if (this._subscription) {
          this._subscription.unsubscribe();
        }
      }

      shouldComponentUpdate(nextProps, nextState) {
        return pure ? !shallowEqual(this.props, nextProps) || !shallowEqual(this.state, nextState) : true;
      }

      render() {
        const props = {
          // state first: connected props are overridable by "actual" (passed) props
          // it's useful for av@queries, av@commands
          ...this.state,
          ...omit(this.props, killProps),
          transition: this.context.transition
        };
        return (
          <Component {...props} />
        );
      }
    };
  };
  if (isKeyList) {
    decorator.Type = { ...pick(stateType.meta.props, select), transition: t.Function };
  }
  return decorator;
};
