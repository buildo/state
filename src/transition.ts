import * as t from 'tcomb';
import shallowEqual from './shallowEqual';
import { BehaviorSubject } from 'rxjs/BehaviorSubject';
import * as debug from 'debug';
import { StateT } from './StateT';

export type TransitionFunctionFunction<S extends StateT> = (s: S) => S;
export type TransitionFunctionPatch<S extends StateT> = { [k in keyof S]?: S[k] | null };
export type TransitionFunction<S extends StateT> = (
  transition: TransitionFunctionFunction<S> | TransitionFunctionPatch<S>
) => void;
export type DryRunTransitionFunction<S extends StateT> = (
  state: S,
  transition: TransitionFunctionFunction<S> | TransitionFunctionPatch<S>
) => {
  patch: TransitionFunctionPatch<S>;
  stateChanged: boolean;
  newState: S;
};

const log = debug('state:transition');

export type MakeTransitionParams<S extends StateT> = {
  stateSubject: BehaviorSubject<S>;
  transitionReducer: TransitionFunctionFunction<S>;
  stateType: t.Interface<S>;
  syncToBrowser: (oldState: S, newState: S) => void;
};
export default function makeTransition<S extends StateT>({
  stateSubject,
  transitionReducer,
  stateType,
  syncToBrowser
}: MakeTransitionParams<S>): {
  transition: TransitionFunction<S>;
  dryRunTransition: DryRunTransitionFunction<S>;
} {
  const dryRunTransition: DryRunTransitionFunction<S> = (state, _transition) => {
    const isTransitionFunction = t.Function.is(_transition);
    const transitionFn = isTransitionFunction
      ? (_transition as TransitionFunctionFunction<S>) // TODO(typo): cast
      : (((() => _transition) as any) as TransitionFunctionFunction<S>); // TODO(typo): cast + removed `t.Object(` check

    const patch = transitionFn(state);

    const shouldReplace = isTransitionFunction;
    const newState = stateType(
      // TODO(typo): double check, it was:
      // transitionReducer(shouldReplace ? { ...patch } : { ...state, ...patch })
      // using Object.assign because of `[ts] Spread types may only be created from object types`
      transitionReducer(shouldReplace ? Object.assign({}, patch) : Object.assign({}, state, patch))
    );

    const stateChanged = !shallowEqual(state, newState);
    return { patch, stateChanged, newState };
  };
  return {
    dryRunTransition,
    transition: _transition => {
      const state = stateSubject.value;
      const { patch, stateChanged, newState } = dryRunTransition(state, _transition);
      log('transition:', _transition);
      log('state is:', state);
      log('patch is:', patch);
      log('new state is:', stateChanged ? '(changed)' : 'NO CHANGE', newState);
      if (stateChanged) {
        stateSubject.next(newState);
        syncToBrowser(state, newState);
      }
    }
  };
}
