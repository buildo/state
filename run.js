import React from 'react';
import debug from 'debug';
import identity from 'lodash/identity';
import find from 'lodash/find';
import { BehaviorSubject } from 'rxjs/subject/BehaviorSubject';
import { ConnectContextTypes } from './connect';
import shallowEqual from './shallowEqual';
import mkContextWrapper from './mkContextWrapper';
import t from 'tcomb';

const log = debug('state:run');

// export for tests
export function createProvideWrapper({
  stateSubject, getPendingState, setPendingState, syncToBrowser, transitionReducer, flushTimeoutMSec
}) {
  const queue = [];
  let lastTransitionId;
  let flushTimeout;

  const maybeRunQueuedTransition = () => {
    if (queue.length > 0) {
      const { args } = queue.shift();
      log('running queued transition', args);
      transition(...args);
    }
  };

  const updateFlushTimeout = () => {
    if (flushTimeout) {
      clearTimeout(flushTimeout);
    }
    flushTimeout = setTimeout(() => {
      if (queue.length > 0 && queue[0].time <= Date.now() - flushTimeoutMSec) {
        setPendingState(null);
        maybeRunQueuedTransition();
        updateFlushTimeout();
      };
    }, flushTimeoutMSec);
  };

  // variadic:
  // (id: String, tr: Tr) => void
  // or
  // (tr: Tr) => void
  //
  // where Tr is `t.Object | t.Function([t.Object], t.Object)`
  const transition = (...args) => {
    const id = args.length === 2 ? t.String(args[0]) : undefined;

    const shouldBeSkipped = !!id && (
      lastTransitionId === id || !!find(queue, ({ args }) => args.length === 2 && args[0] === id)
    );

    if (shouldBeSkipped) {
      log(`transition: skipping transition '${id}'`);
      return;
    }

    if (getPendingState()) {
      log('transition: enqueing this transition:', args);
      queue.push({ args, time: Date.now() });
      updateFlushTimeout();
      // log('transition queue:', queue);
      return;
    }

    log('transition:', args);
    lastTransitionId = id;

    const _transitionFn = args.length === 2 ? args[1] : args[0];
    const isTransitionFunction = t.Function.is(_transitionFn);
    const transitionFn = isTransitionFunction ? _transitionFn : () => t.Object(_transitionFn);

    const state = stateSubject.value;
    const patch = transitionFn(state);

    // log('state is:', state);
    // log('patch is:', patch);
    const shouldReplace = isTransitionFunction;
    const newState = transitionReducer(shouldReplace ? { ...patch } : { ...state, ...patch });

    for (const k in patch) { // eslint-disable-line no-loops/no-loops
      if (t.Nil.is(patch[k])) {
        delete newState[k];
      }
    }

    const stateChanged = !shallowEqual(state, newState);
    // log('new state is:', newState, stateChanged ? '(changed)' : 'NO CHANGE');
    if (stateChanged) {
      const wroteToRouter = syncToBrowser(state, newState);
      if (!wroteToRouter) {
        // log(`didn't write to router, nexting newState`);
        stateSubject.next(newState);
      } else {
        // log(`wrote to router, setting pendingState`);
        setPendingState(newState);
      }
    } else {
      maybeRunQueuedTransition();
    }
  };

  class ProvideWrapper extends React.Component {
    static childContextTypes = ConnectContextTypes

    getChildContext = () => ({
      transition, state: stateSubject
    })

    render() {
      return this.props.children();
    }
  }

  // this subscription is (one of) the first
  // i.e. every connected component will receive updates after this
  // ...meaning decisions taken before `transition` time become risky
  // in presence of other interleaving transitions
  //
  // you can always make choices at `transition` time
  // using the `transition(State => State)` form
  stateSubject.subscribe(maybeRunQueuedTransition);

  return { ProvideWrapper, transition };
}

export default function run({
  // final render
  // could be something like:
  // element => {
  //   React.render(element, mountNode);
  // }
  //
  // (element: ReactElement) => void
  //
  render, // required

  // sync a change to the browser
  // should return true if a change to browser was pushed
  //
  // (oldState: Object, newState: Object) => wrote: Boolean
  //
  syncToBrowser, // required

  // let me subscribe to browser changes
  // should invoke callback at every broser change
  // with the (possibly new) component to be rendered
  // and the (possibly new) state from browser
  // new state
  //
  // (callback:
  //   (component: ReactNode, newStateFromBrowser: Object) => void
  // ) => void
  //
  onBrowserChange, // required

  // provide this custom context object to the tree
  // see also provideContextTypes
  //
  // Object
  //
  provideContext = {},

  // provide this custom contextTypes to the tree
  // see also provideContext
  //
  // Object
  //
  provideContextTypes = {},

  // initial state
  //
  // Object
  //
  initialState = {},

  // apply custom transformations after every state transition
  //
  // (state: Object) => Object
  //
  transitionReducer = identity,

  // merge current state and new browser state
  //
  // (state: Object, browserState: Object) => Object
  //
  mergeStateAndBrowserState = (s, b) => ({ ...s, ...b }),

  // subscribe to every state change
  // ...in case you need it
  //
  // (state: Object) => void
  //
  subscribe = () => {},

  // app own initialization calback
  // ...in case you need it
  // has access access the state Rx.Subject and
  // the transition function (documented elsewhere)
  //
  // (state: Rx.Subject, transition: Function) => void
  //
  init = () => {},

  // force next transition even if pending state
  // is not yet applied after `flushTimeoutMsec`
  //
  // Number
  //
  flushTimeoutMSec = 200
}) {
  const state = new BehaviorSubject(initialState);
  state.subscribe(subscribe);
  let _newState;

  const { ProvideWrapper, transition } = createProvideWrapper({
    stateSubject: state,
    syncToBrowser,
    getPendingState: () => _newState,
    setPendingState: s => { _newState = s; },
    transitionReducer,
    flushTimeoutMSec
  });


  const Provider = mkContextWrapper(provideContext, provideContextTypes);

  onBrowserChange((renderElement: t.Function, fromRouter: t.Object) => {
    if (_newState) {
      log('browser change: user initiated');
      const newState = mergeStateAndBrowserState(_newState, fromRouter);
      _newState = null;
      state.next(newState);
    } else {
      log('browser change: browser initiated');
      state.next(mergeStateAndBrowserState(state.value, fromRouter));
    }
    /* eslint-disable react/display-name */
    render(
      <ProvideWrapper>
        {() => <Provider>{renderElement}</Provider>}
      </ProvideWrapper>
    );
    /* eslint-enable react/display-name */
  });

  transition('start', initialState);

  init(state, transition);

  // return for tests
  return transition;
}
