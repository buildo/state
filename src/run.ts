import * as debug from 'debug';
import difference = require('lodash/difference');
import pickBy = require('lodash/pickBy');
import omitBy = require('lodash/omitBy');
import { BehaviorSubject } from 'rxjs/BehaviorSubject';
import { ConnectContextTypes } from './connect';
import mkContextWrapper from './mkContextWrapper';
import mkTransition from './transition';
import { TransitionFunction, TransitionFunctionFunction } from './transition';
import * as t from 'tcomb';
import { syncToBrowser, onBrowserChange, BrowserState } from './browser';
import { StateT } from './StateT';

declare var process: any; // TODO(typo)

const log = debug('state:run');

const omitNils = <S extends StateT>(s: S) => omitBy<S, S>(s, t.Nil.is);

export type StateContextWrapper = React.ComponentType<{}>;

type StateSubject<S extends StateT> = BehaviorSubject<S>;
type StateTcombType<S extends StateT> = t.Interface<S>;

type RunParams<S extends StateT> = {
  initialState: S;
  transitionReducer?: TransitionFunctionFunction<S>;
  subscribe?: (s: S) => void;
  init?: (s: StateSubject<S>, t: TransitionFunction<S>) => void;
  shouldSerializeKey?: (k: keyof S) => boolean;
  shouldBrowserPatchBePushedOrReplaced?: (oldState: S, newState: S) => boolean;
};
type RunReturn = Promise<StateContextWrapper>;
export type Run<S extends StateT> = (p: RunParams<S>) => RunReturn;

function pickValidStateKeys<S extends StateT>(stateType: StateTcombType<S>, b: BrowserState): S {
  return pickBy<S, BrowserState>(b, (v, k) => {
    return stateType.meta.props.hasOwnProperty(k) && (stateType.meta.props[k] as t.Type<any>).is(v);
  });
}

const mergeStateAndBrowserState = <S extends StateT>(stateType: StateTcombType<S>) => (s: S, b: BrowserState) =>
  omitNils({ ...s as any, ...pickValidStateKeys(stateType, b) as any }) as S; // TODO(typo)

const transitionReducerIdentity = <S extends StateT>(s: S) => s;

export default <S extends StateT>(stateType: StateTcombType<S>) => ({
  // initial state
  initialState,

  // apply custom transformations after every state transition
  transitionReducer: _transitionReducer = transitionReducerIdentity,

  // subscribe to every state change
  // ...in case you need it
  subscribe = () => {},

  // app own initialization calback
  // ...in case you need it
  // has access access the state Rx.Subject and
  // the transition function (documented elsewhere)
  init = () => {},

  // whether to serialize (sync to browser) a state key or not
  shouldSerializeKey = () => true,

  // whether a patch should be serialized (synced to browser) as a new history item
  // (true and default) or replace the current history item (false)
  shouldBrowserPatchBePushedOrReplaced = () => true
}: RunParams<S>): RunReturn =>
  new Promise(resolve => {
    const transitionReducer: TransitionFunctionFunction<S> = (s: S) => omitNils<S>(_transitionReducer(s));

    const state = new BehaviorSubject(stateType(initialState));
    state.subscribe(subscribe);

    const transition = mkTransition({
      stateSubject: state,
      stateType,
      syncToBrowser: (oldState, newState) => {
        const newSerialized = pickBy<BrowserState, BrowserState>(newState, (_: any, k: keyof S) =>
          shouldSerializeKey(k)
        );
        if (process.env.NODE_ENV === 'development') {
          log('syncing to browser, omitted:', difference(Object.keys(newState), Object.keys(newSerialized)));
        }
        return syncToBrowser(newSerialized, shouldBrowserPatchBePushedOrReplaced(oldState, newState));
      },
      transitionReducer
    });

    const ProvideWrapper = mkContextWrapper({ transition, state }, ConnectContextTypes);

    // wait to receive the first browser state before resolving, so that users can
    // render with something meaningful at hand
    let _bootstrapped = false;

    const mergeStateAndValidBrowserState = mergeStateAndBrowserState<S>(stateType);

    onBrowserChange(fromRouter => {
      log('browser change');
      const newState = mergeStateAndValidBrowserState(state.value, fromRouter);
      transition(newState);

      if (!_bootstrapped) {
        _bootstrapped = true;
        resolve(ProvideWrapper);
      }
    });

    init(state, transition);
  });
