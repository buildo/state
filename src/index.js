import run from './run';
import connect from './connect';
import t from 'tcomb';

const isStrictInterface = (x) => {
  return t.isType(x) && (x.meta.kind === 'interface') && (x.meta.strict === true);
};

export default stateType => {
  if (!isStrictInterface(stateType)) {
    throw new Error('`stateType` must be a strict tcomb interface');
  }

  return {
    run: run(stateType),
    connect: connect(stateType)
  };
};
