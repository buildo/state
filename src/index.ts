import * as t from 'tcomb';
import run, { Run } from './run';
import connect, { Connect } from './connect';
import { StateT } from './StateT';

const isStrictInterface = (x: any) => {
  return t.isType(x) && x.meta.kind === 'interface' && x.meta.strict === true;
};

export default function<S extends StateT>(
  _stateType: t.Interface<S>
): {
  run: Run<S>;
  appState: t.Interface<S>;
  connect: Connect<S>;
} {
  if (!isStrictInterface(_stateType)) {
    throw new Error('`stateType` must be a strict tcomb interface');
  }

  const stateType = t.interface(
    {
      ___k: t.maybe(t.Number),
      ..._stateType.meta.props
    },
    { strict: true, name: _stateType.meta.name }
  ) as t.Interface<S>;

  return {
    appState: stateType,
    run: args => run(stateType)(args),
    connect: connect(stateType)
  };
}
