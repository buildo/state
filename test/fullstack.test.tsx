import make from '../src';
import { RunParams } from '../src/run';
import * as t from 'tcomb';
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { TransitionFunction } from '../src/transition';
import createMemoryHistory from 'history/createMemoryHistory';

type State = {
  view: 'view1' | 'view2';
  foo?: string;
  bar?: number;
};
const State = t.interface<State>(
  {
    view: t.enums.of(['view1', 'view2']),
    foo: t.maybe(t.String),
    bar: t.maybe(t.Number)
  },
  { strict: true }
);

class RenderState extends React.Component<Partial<State>> {
  render() {
    return (
      <div>
        {JSON.stringify(this.props)}
      </div>
    );
  }
}

const snapshot = (Provider, App) =>
  renderer
    .create(
      <Provider>
        {() => <App />}
      </Provider>
    )
    .toJSON();

type InitialLocationEntry = string | { pathname: string; search?: string };
const simpleScenario = (
  {
    initialEntries = [],
    transitionReducer
  }: { initialEntries?: InitialLocationEntry[]; transitionReducer?: RunParams<State>['transitionReducer'] } = {}
) => {
  const { run, connect } = make<State>(State);
  const App = connect(['view', 'foo', 'bar'])(RenderState);
  let transition: TransitionFunction<State>;
  const states: State[] = [];
  const history = createMemoryHistory({ initialEntries } as { initialEntries: string[] }); // history typing is slightly broken here
  return run({
    initialState: { view: 'view1' },
    history,
    transitionReducer,
    init: (_: any, tr) => {
      transition = tr;
    },
    subscribe: s => {
      states.push(s);
    }
  }).then(Provider => ({
    transition,
    states,
    history,
    snapshot: () => snapshot(Provider, App)
  }));
};

describe('fullstack', () => {
  it('updating state with something valid should work', () =>
    new Promise((resolve, reject) => {
      return simpleScenario()
        .then(({ transition, states, history, snapshot }) => {
          transition({ view: 'view2', bar: 4 });
          expect(history.location.pathname).toBe('/view2');
          expect(history.location.search).toBe('?bar=4');
          expect(history.length).toBe(1);
          expect(states[states.length - 1]).toEqual({ view: 'view2', bar: 4 });
          expect(snapshot()).toMatchSnapshot();
        })
        .then(() => resolve(), reject);
    }));

  it('updating state multiple times with same value should cause a single diff and push', () =>
    new Promise((resolve, reject) => {
      return simpleScenario()
        .then(({ transition, states, history, snapshot }) => {
          transition({ view: 'view2', bar: 4 });
          transition({ view: 'view2', bar: 4 });
          transition({ view: 'view2', bar: 4 });
          expect(history.location.pathname).toBe('/view2');
          expect(history.location.search).toBe('?bar=4');
          expect(history.length).toBe(1);
          expect(states[states.length - 1]).toEqual({ view: 'view2', bar: 4 });
          expect(snapshot()).toMatchSnapshot();
        })
        .then(() => resolve(), reject);
    }));

  it('updating state 3 times with changes should cause 3 diffs and pushes', () =>
    new Promise((resolve, reject) => {
      return simpleScenario()
        .then(({ transition, states, history, snapshot }) => {
          transition({ view: 'view2', foo: 'bar' });
          expect(history.location.pathname).toBe('/view2');
          expect(history.location.search).toBe('?foo=bar');
          expect(history.length).toBe(1);
          expect(states[states.length - 1]).toEqual({ view: 'view2', foo: 'bar' });
          expect(snapshot()).toMatchSnapshot();

          transition({ view: 'view2', foo: 'baz' });
          expect(history.location.pathname).toBe('/view2');
          expect(history.location.search).toBe('?foo=baz');
          expect(history.length).toBe(2);
          expect(states[states.length - 1]).toEqual({ view: 'view2', foo: 'baz' });
          expect(snapshot()).toMatchSnapshot();

          transition({ view: 'view2', foo: 'bar' });
          expect(history.location.pathname).toBe('/view2');
          expect(history.location.search).toBe('?foo=bar');
          expect(history.length).toBe(3);
          expect(states[states.length - 1]).toEqual({ view: 'view2', foo: 'bar' });
          expect(snapshot()).toMatchSnapshot();

          expect(states.length).toBe(4);
        })
        .then(() => resolve(), reject);
    }));

  it('should pick up from browser correctly', () =>
    new Promise((resolve, reject) => {
      const initialEntries = [{ pathname: '/view2', search: '?bar=1', hash: '' }];
      return simpleScenario({ initialEntries })
        .then(({ transition, states, history, snapshot }) => {
          expect(history.location.pathname).toBe('/view2');
          expect(history.location.search).toBe('?bar=1');
          // expect(history.length).toBe(1); // TODO(gio): this fails, probably wouldn't fail in a real browser :(
          expect(states.length).toBe(1 + 1); // initial state + picked up state
          expect(states[states.length - 1]).toEqual({ view: 'view2', bar: 1 });
          expect(snapshot()).toMatchSnapshot();
        })
        .then(() => resolve(), reject);
    }));

  it('should pick up from browser correctly with invalid params', () =>
    new Promise((resolve, reject) => {
      const initialEntries = [{ pathname: '/view2', search: '?bar=false', hash: '' }];
      return simpleScenario({ initialEntries })
        .then(({ transition, states, history, snapshot }) => {
          expect(history.location.pathname).toBe('/view2');
          expect(history.location.search).toBe('?');
          // expect(history.length).toBe(1); // TODO(gio): this fails, probably wouldn't fail in a real browser :(
          expect(states.length).toBe(1 + 1); // initial state + picked up state
          expect(states[states.length - 1]).toEqual({ view: 'view2' });
          expect(snapshot()).toMatchSnapshot();
        })
        .then(() => resolve(), reject);
    }));

  it('updating state with something valid and serialized should work', () =>
    new Promise((resolve, reject) => {
      return simpleScenario()
        .then(({ transition, states, history, snapshot }) => {
          transition({ bar: 2 });
          expect(history.location.pathname).toBe('/view1');
          expect(history.location.search).toBe('?bar=2');
          expect(history.length).toBe(1);
          expect(states.length).toBe(2);
          expect(states[states.length - 1]).toEqual({ view: 'view1', bar: 2 });
          expect(snapshot()).toMatchSnapshot();
        })
        .then(() => resolve(), reject);
    }));

  it('should react correctly to browser back + fwd', () =>
    new Promise((resolve, reject) => {
      return simpleScenario()
        .then(({ transition, states, history, snapshot }) => {
          // add two entries to history
          transition({ bar: 2, view: 'view2' });
          transition({ bar: 4, view: 'view1' });
          expect(history.location.pathname).toBe('/view1');
          expect(history.location.search).toBe('?bar=4');
          expect(history.length).toBe(2);
          expect(history.index).toBe(1);
          expect(states.length).toBe(3);
          expect(states[states.length - 1]).toEqual({ view: 'view1', bar: 4 });
          expect(snapshot()).toMatchSnapshot();
          // browser back
          history.goBack();
          expect(history.location.pathname).toBe('/view2');
          expect(history.location.search).toBe('?bar=2');
          expect(history.length).toBe(2);
          expect(history.index).toBe(0);
          expect(states.length).toBe(4);
          expect(states[states.length - 1]).toEqual({ view: 'view2', bar: 2 });
          expect(snapshot()).toMatchSnapshot();
          // browser fwd
          history.goForward();
          expect(history.location.pathname).toBe('/view1');
          expect(history.location.search).toBe('?bar=4');
          expect(history.length).toBe(2);
          expect(history.index).toBe(1);
          expect(states.length).toBe(5);
          expect(states[states.length - 1]).toEqual({ view: 'view1', bar: 4 });
          expect(snapshot()).toMatchSnapshot();
        })
        .then(() => resolve(), reject);
    }));

  it('should react correctly to a browser back + redirect', () =>
    new Promise((resolve, reject) => {
      let redirect = false;
      return simpleScenario({
        transitionReducer: s => ({ ...s, view: redirect ? 'view1' : s.view })
      })
        .then(({ transition, states, history, snapshot }) => {
          // add two entries to history
          transition({ bar: 2, view: 'view2' });
          transition({ bar: 4, view: 'view1' });
          expect(snapshot()).toMatchSnapshot();
          // browser back forcing a redirect
          redirect = true;
          history.goBack();
          expect(history.location.pathname).toBe('/view1');
          expect(history.location.search).toBe('?bar=2');
          expect(history.length).toBe(2);
          expect(history.index).toBe(0);
          expect(states.length).toBe(4);
          expect(states[states.length - 1]).toEqual({ view: 'view1', bar: 2 });
          expect(snapshot()).toMatchSnapshot();
        })
        .then(() => resolve(), reject);
    }));
});
