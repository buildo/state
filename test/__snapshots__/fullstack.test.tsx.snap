// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fullstack should pick up from browser correctly 1`] = `
<div>
  {"view":"view2","bar":1}
</div>
`;

exports[`fullstack should pick up from browser correctly with invalid params 1`] = `
<div>
  {"view":"view2"}
</div>
`;

exports[`fullstack should react correctly to a browser back + redirect 1`] = `
<div>
  {"view":"view1","bar":4}
</div>
`;

exports[`fullstack should react correctly to browser back + fwd 1`] = `
<div>
  {"view":"view1","bar":4}
</div>
`;

exports[`fullstack should react correctly to browser back + fwd 2`] = `
<div>
  {"view":"view2","bar":2}
</div>
`;

exports[`fullstack should react correctly to browser back + fwd 3`] = `
<div>
  {"view":"view1","bar":4}
</div>
`;

exports[`fullstack updating state 3 times with changes should cause 3 diffs and pushes 1`] = `
<div>
  {"view":"view2","foo":"bar"}
</div>
`;

exports[`fullstack updating state 3 times with changes should cause 3 diffs and pushes 2`] = `
<div>
  {"view":"view2","foo":"baz"}
</div>
`;

exports[`fullstack updating state 3 times with changes should cause 3 diffs and pushes 3`] = `
<div>
  {"view":"view2","foo":"bar"}
</div>
`;

exports[`fullstack updating state multiple times with same value should cause a single diff and push 1`] = `
<div>
  {"view":"view2","bar":4}
</div>
`;

exports[`fullstack updating state with something valid and serialized should work 1`] = `
<div>
  {"view":"view1","bar":2}
</div>
`;

exports[`fullstack updating state with something valid should work 1`] = `
<div>
  {"view":"view2","bar":4}
</div>
`;
